llarp/link/link_manager.cpp

```diff
@@ -1062,14 +1062,178 @@ namespace llarp
   void
   LinkManager::handle_path_build(oxen::quic::message m)
   {
+    if (!router.path_context().AllowingTransit())
+    {
+      log::warning("got path build request when not permitting transit");
+      m.respond(serialize_response({{"STATUS", PathBuildMessage::NO_TRANSIT}}), true);
+      return;
+    }
     try
     {
-      oxenc::bt_dict_consumer btdc{m.body()};
+      // not using list_consumer here because we need to move the first (our) frame
+      // to the end and re-send, unless we're the terminal hop.  This could be done
+      // with list_consumer, but it would involve messy mucking around with the
+      // encoded list.  Revisit if optimization issue (shouldn't be).
+      auto frame_list = oxenc::bt_deserialize<oxenc::bt_list>(m.body());
+      if (frames.size() != path::MAX_LEN)
+      {
+        log::info(
+            link_cat,
+            "Path build request has invalid number of records, ",
+            frame_list.size(),
+            "!=",
+            path::MAX_LEN);
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_FRAMES}}), true);
+        return;
+      }
+
+      const auto& hash = frame_list[0].at("hash");
+      auto& frame_str = frame_list[0].at("frame");
+
+      oxenc::bt_dict_consumer frame_dict{frame_str};
+
+      auto hop_info = frame_dict.require<ustring>("encrypted");
+      TunnelNonce nonce;
+      nonce.from_string_view(frame_dict.require<std::string_view>("nonce"));
+      Pubkey otherPubkey;
+      otherPubkey.from_string_view(frame_dict.require<std::string_view>("pubkey"));
+
+      auto crypto = CryptoManager::instance();
+
+      SharedSecret shared;
+      // derive shared secret using ephemeral pubkey and our secret key (and nonce)
+      if (!crypto->dh_server(shared, otherPubkey, router.identity(), nonce))
+      {
+        log::info("DH failed during path build.");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+
+      // hash data and check against given hash
+      ShortHash digest;
+      if (!crypto->hmac(
+              digest.data(),
+              reinterpret_cast<unsigned char*>(frame_str.data()),
+              frame_str.size(),
+              shared))
+      {
+        log::error(link_cat, "HMAC failed on path build request");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+      if (!std::equal(
+              digest.begin(), digest.end(), reinterpret_cast<const unsigned char*>(hash.data())))
+      {
+        log::info("HMAC mismatch on path build request");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+
+      // decrypt frame with our hop info
+      if (!crypto->xchacha20(
+              reinterpret_cast<unsigned char*>(hop_info.data()), hop_info.size(), shared, nonce))
```

Comment:
This is already a ustring so luckily we dont need to reinterpret cast (since Jason's oxenc PR)